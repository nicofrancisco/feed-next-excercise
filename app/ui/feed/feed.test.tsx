// components/__tests__/Feed.test.tsx


import React from 'react'
import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { vi, expect } from 'vitest'
import Feed from './feed'
import { Post, User } from '../../lib/types'
import configureMockStore from 'redux-mock-store'
import { debug } from 'console'

const mockStore = configureMockStore()
/*vi.mock('redux-mock-store', () => {
  const actual = vi.importActual('redux-mock-store')
  return {
    ...actual,
    default: () => {
      const mockStore = (state: any) => ({
        getState: () => state,
        subscribe: vi.fn(),
        dispatch: vi.fn(),
        clearActions: vi.fn(), // Add clearActions if needed
        getActions: vi.fn(() => []), // Add getActions if needed
      })
      return mockStore
    },
  }
})
const mockStore = require('redux-mock-store').default*/

const mockPosts: Record<number, Post> = {
  1: { id: 1, title: 'Test Post 1', body: 'This is the body of test post 1', userId: 1 },
  2: { id: 2, title: 'Test Post 2', body: 'This is the body of test post 2', userId: 2 },
}
const mockUsers: Record<number, User> = {
  1: { 
    id: 1, 
    name: 'Test User 1', 
    username: 'testuser1', 
    email: 'test1@example.com', 
    phone: '123-456-9990', 
    website: 'www.user1.com',
    address: { street: '123 Test St', city: 'Test', zipcode: '12345', suite:'', geo:{
      lat: '',
      lng: ''}
    }, 
    company: { name: 'Test Company 1', catchPhrase: 'Test 1', bs: 'Test as 1' } // Add company
  },
  2: { 
    id: 2, 
    name: 'Test User 2', 
    username: 'testuser2', 
    email: 'test2@example.com', 
    phone: '123-456-9991', 
    website: 'www.user2.com',
    address: { street: '123 Test St', city: 'Test', zipcode: '12345', suite:'', geo:{
      lat: '',
      lng: ''}
    },
    company: { name: 'Test Company 2', catchPhrase: 'Test 2', bs: 'Test as 2' } // Add company
  },
}

const mockParams = {
  displayedPostIds: [1, 2],
  posts: mockPosts,
  newPosts: {
    ids: [1, 2],
    entities: mockPosts,
  },
  users: mockUsers,
  scrollPosition: 0,
  onScroll: vi.fn(),
  fetchMorePost: vi.fn(),
}

describe('Feed Component', () => {
  it('renders correctly', () => {
    const initialState = {
      // Add your initial state here if needed
    }
    const store = mockStore(initialState)

    const { asFragment } = render(
      <Provider store={store}>
        <Feed {...mockParams} />
      </Provider>
    )

    debug()

    // Check if the user names are present
    expect(screen.queryAllByText('Test User 1')).toBeTruthy()
    expect(screen.queryAllByText('Test User 2')).toBeTruthy()

    // Check if the post titles are present
    expect(screen.queryAllByText('Test Post 1')).toBeTruthy()
    expect(screen.queryAllByText('Test Post 2')).toBeTruthy()

    // Snapshot test for the entire component
    expect(asFragment()).toMatchSnapshot()
  })
})

/*import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import Feed from './feed';
import { vi, expect } from 'vitest'

vi.mock('redux-mock-store', () => {
  const actual = vi.importActual('redux-mock-store')
  return {
    ...actual,
    default: () => {
      const mockStore = (state:any) => ({
        getState: () => state,
        subscribe: vi.fn(),
        dispatch: vi.fn(),
      })
      return mockStore
    },
  }
})
const mockStore = require('redux-mock-store').default

describe('Feed Component', () => {
  it('renders correctly', () => {
    const initialState = {
      posts: {
        items: [
          { id: 1, title: 'Test Post', body: 'This is a test post', user: { name: 'Test User' } },
        ],
        status: 'succeeded',
      },
    };
    const store = mockStore(initialState);

    const { asFragment } = render(
      <Provider store={store}>
        <Feed displayedPostIds={[]} posts={[]} newPosts={[]} users={[]} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
})*/

class IntersectionObserver {
  root: Element | null = null;
  rootMargin: string = '';
  thresholds: ReadonlyArray<number> = [];
  
  constructor(callback: IntersectionObserverCallback, options?: IntersectionObserverInit) {}
  observe() {}
  unobserve() {}
  disconnect() {}
  takeRecords() { return []; }
}

global.IntersectionObserver = IntersectionObserver