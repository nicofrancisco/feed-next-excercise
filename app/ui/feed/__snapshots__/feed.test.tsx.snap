// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Feed Component > renders correctly 1`] = `
<DocumentFragment>
  <div
    class="grid grid-rows-12 h-screen animate-fadeInUp flex-col"
  >
    <ul
      class="py-5 row-start-2 row-span-10 overflow-y-auto sm:row-start-1 sm:row-span-11"
      role="list"
    >
      <div
        data-testid="post-preview"
      >
        <a
          aria-label="Read post: "Test Post 1" by Test User 1"
          class=" px-5 py-2 flex justify-between hover:bg-gray-100 sm:px-7"
          href="/post?id=1"
        >
          <article
            class=" flex space-x-4"
          >
            <div
              class=" flex space-x-4"
            >
              <div
                class="min-w-min flex flex-col items-center row-start-1 row-end-7 col-start-1 col-end-3"
              >
                <img
                  alt="user-avatar"
                  class="min-h-14 min-w-14 flex-none rounded-full bg-gray-50"
                  loading="lazy"
                  src="https://api.dicebear.com/8.x/adventurer/svg?seed=Test User 1"
                />
                <div
                  class="grow border-l-4 border-slate-200 min-h-5 w-0 mt-2"
                />
              </div>
              <div
                class="flex flex-col"
              >
                <div
                  class="h-16 flex flex-col justify-center"
                >
                  <span
                    class="font-bold text-lg"
                  >
                    Test User 1
                  </span>
                  <time
                    class="text-xs leading-5 text-gray-500"
                  >
                    Old record
                  </time>
                </div>
                <h2
                  class="text-md font-semibold leading-6 text-gray-700 capitalize"
                >
                  Test Post 1
                </h2>
                <p
                  class="mt-2 line-clamp-3 max-h-32 text-wrap overflow-y-hidden text-ellipsis overflow-hidden text-sm leading-6 text-gray-800"
                >
                  This is the body of test post 1
                </p>
              </div>
            </div>
          </article>
        </a>
      </div>
      <div
        data-testid="post-preview"
      >
        <a
          aria-label="Read post: "Test Post 2" by Test User 2"
          class=" px-5 py-2 flex justify-between hover:bg-gray-100 sm:px-7"
          href="/post?id=2"
        >
          <article
            class=" flex space-x-4"
          >
            <div
              class=" flex space-x-4"
            >
              <div
                class="min-w-min flex flex-col items-center row-start-1 row-end-7 col-start-1 col-end-3"
              >
                <img
                  alt="user-avatar"
                  class="min-h-14 min-w-14 flex-none rounded-full bg-gray-50"
                  loading="lazy"
                  src="https://api.dicebear.com/8.x/adventurer/svg?seed=Test User 2"
                />
                <div
                  class="grow border-l-4 border-slate-200 min-h-5 w-0 mt-2"
                />
              </div>
              <div
                class="flex flex-col"
              >
                <div
                  class="h-16 flex flex-col justify-center"
                >
                  <span
                    class="font-bold text-lg"
                  >
                    Test User 2
                  </span>
                  <time
                    class="text-xs leading-5 text-gray-500"
                  >
                    Old record
                  </time>
                </div>
                <h2
                  class="text-md font-semibold leading-6 text-gray-700 capitalize"
                >
                  Test Post 2
                </h2>
                <p
                  class="mt-2 line-clamp-3 max-h-32 text-wrap overflow-y-hidden text-ellipsis overflow-hidden text-sm leading-6 text-gray-800"
                >
                  This is the body of test post 2
                </p>
              </div>
            </div>
          </article>
        </a>
      </div>
      <div
        data-testid="post-preview"
      >
        <a
          aria-label="Read post: "Test Post 1" by Test User 1"
          class=" px-5 py-2 flex justify-between hover:bg-gray-100 sm:px-7"
          href="/post?id=1"
        >
          <article
            class=" flex space-x-4"
          >
            <div
              class=" flex space-x-4"
            >
              <div
                class="min-w-min flex flex-col items-center row-start-1 row-end-7 col-start-1 col-end-3"
              >
                <img
                  alt="user-avatar"
                  class="min-h-14 min-w-14 flex-none rounded-full bg-gray-50"
                  loading="lazy"
                  src="https://api.dicebear.com/8.x/adventurer/svg?seed=Test User 1"
                />
                <div
                  class="grow border-l-4 border-slate-200 min-h-5 w-0 mt-2"
                />
              </div>
              <div
                class="flex flex-col"
              >
                <div
                  class="h-16 flex flex-col justify-center"
                >
                  <span
                    class="font-bold text-lg"
                  >
                    Test User 1
                  </span>
                  <time
                    class="text-xs leading-5 text-gray-500"
                  >
                    Old record
                  </time>
                </div>
                <h2
                  class="text-md font-semibold leading-6 text-gray-700 capitalize"
                >
                  Test Post 1
                </h2>
                <p
                  class="mt-2 line-clamp-3 max-h-32 text-wrap overflow-y-hidden text-ellipsis overflow-hidden text-sm leading-6 text-gray-800"
                >
                  This is the body of test post 1
                </p>
              </div>
            </div>
          </article>
        </a>
      </div>
      <div
        data-testid="post-preview"
      >
        <a
          aria-label="Read post: "Test Post 2" by Test User 2"
          class=" px-5 py-2 flex justify-between hover:bg-gray-100 sm:px-7"
          href="/post?id=2"
        >
          <article
            class=" flex space-x-4"
          >
            <div
              class=" flex space-x-4"
            >
              <div
                class="min-w-min flex flex-col items-center row-start-1 row-end-7 col-start-1 col-end-3"
              >
                <img
                  alt="user-avatar"
                  class="min-h-14 min-w-14 flex-none rounded-full bg-gray-50"
                  loading="lazy"
                  src="https://api.dicebear.com/8.x/adventurer/svg?seed=Test User 2"
                />
                <div
                  class="grow border-l-4 border-slate-200 min-h-5 w-0 mt-2"
                />
              </div>
              <div
                class="flex flex-col"
              >
                <div
                  class="h-16 flex flex-col justify-center"
                >
                  <span
                    class="font-bold text-lg"
                  >
                    Test User 2
                  </span>
                  <time
                    class="text-xs leading-5 text-gray-500"
                  >
                    Old record
                  </time>
                </div>
                <h2
                  class="text-md font-semibold leading-6 text-gray-700 capitalize"
                >
                  Test Post 2
                </h2>
                <p
                  class="mt-2 line-clamp-3 max-h-32 text-wrap overflow-y-hidden text-ellipsis overflow-hidden text-sm leading-6 text-gray-800"
                >
                  This is the body of test post 2
                </p>
              </div>
            </div>
          </article>
        </a>
      </div>
      <div
        class="flex justify-center mt-2"
      />
    </ul>
  </div>
</DocumentFragment>
`;
